common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java

```diff
@@ -0,0 +1,249 @@
+    }
+
+    @VisibleForTesting
+    String getWatchedFileName() {
+        Path fileName = this.watchedFilePath.getFileName();
+        if (fileName != null) {
+            return fileName.toString();
+        } else {
+            // It should never happen that file name is null, but we have to perform this defensive null check to
+            // satisfy SpotBugs. File name will be null for a path like "/".
+            throw new IllegalStateException("File name is null");
+        }
+    }
+
+    @VisibleForTesting
+    Path getWatchedDirectory() {
+        return this.watchedFilePath.getParent();
+    }
+
+    @Override
+    @SuppressWarnings("SleepWhileInLoop")
+    public void run() {
+        WatchKey watchKey = null;
+        WatchService watchService = null;
+        try {
+            watchService = FileSystems.getDefault().newWatchService();
+            log.debug("Done creating watch service for watching file at path: {}", this.watchedFilePath);
+
+            String fileName = getWatchedFileName();
+            Path directoryPath = getWatchedDirectory();
+            log.debug("Directory being watched is {}", directoryPath);
+
+            assert directoryPath != null;
+            directoryPath.register(watchService, StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.ENTRY_CREATE);
+            log.debug("Done setting up watch for modify entries for file at path: {}", this.watchedFilePath);
+
+            isWatchRegistered = true;
+
+            while (!Thread.currentThread().isInterrupted()) {
+                try {
+                    watchKey = retrieveWatchKeyFrom(watchService);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    logException(e);
+                    // Allow thread to exit.
+                }
+
+                if (watchKey != null) {
+                    Optional<WatchEvent<?>> modificationDetectionEvent = watchKey.pollEvents()
+                            .stream()
+                            .filter( // we only care about changes to the specified file.
+                                    event -> event.context().toString().contains(fileName))
+                            .findAny();
+
+                    if (modificationDetectionEvent.isPresent()) {
+                        log.info("Detected that the file [{}] has modified", this.watchedFilePath);
+                        callback.accept(modificationDetectionEvent.get());
+                    }
+
+                    boolean isKeyValid = watchKey.reset();
+                    log.debug("Done resetting watch key.");
+                    if (!isKeyValid) {
+                        log.info("No longer watching file [{}]", this.watchedFilePath);
+                        break;
+                    }
+                } else {
+                    log.debug("watchKey for file at path {} was null", this.watchedFilePath);
```

Comment:
If retrieveWatchKeyFrom is interrupted, then the catch block will allow the iteration to proceed and execute this log statement, which might not be meaningful in the case of an interruption. I'm essentially wondering whether we should be skipping the iteration altogether in the case of an interrupt. Maybe it is harmless, though.

Reply:
Sure, I have removed the log statement.